src/cls/cmpomap/server.cc

```diff
@@ -0,0 +1,294 @@
+    if (r == -EIO) {
+      r = 0; // treat EIO as a failed comparison
+    }
+    if (r < 0) {
+      CLS_LOG(10, "cmp_set_vals() failed to compare key=%s r=%d",
+              key.c_str(), r);
+      return r;
+    }
+    if (r == 0) {
+      // unsuccessful comparison
+      if (k != values.end()) {
+        values.erase(k); // remove this key from the values to overwrite
+        CLS_LOG(20, "cmp_set_vals() not overwriting key=%s", key.c_str());
+      } else {
+        CLS_LOG(20, "cmp_set_vals() not writing missing key=%s", key.c_str());
+      }
+    } else {
+      // successful comparison
+      if (k != values.end()) {
+        // overwrite the value
+        k->second = std::move(cmp.value);
+        CLS_LOG(20, "cmp_set_vals() overwriting key=%s", key.c_str());
+      } else {
+        // insert the value
+        values.emplace(key, std::move(cmp.value));
+        CLS_LOG(20, "cmp_set_vals() overwriting missing key=%s", key.c_str());
+      }
+    }
+  }
+
+  if (values.empty()) {
+    CLS_LOG(20, "cmp_set_vals() has no values to overwrite");
+    return 0;
+  }
+
+  CLS_LOG(20, "cmp_set_vals() overwriting count=%d", (int)values.size());
+  return cls_cxx_map_set_vals(hctx, &values);
+}
+
+static int cmp_rm_keys(cls_method_context_t hctx, bufferlist *in, bufferlist *out)
+{
+  cmp_rm_keys_op op;
+  try {
+    auto p = in->cbegin();
+    decode(op, p);
+  } catch (const buffer::error&) {
+    CLS_LOG(1, "ERROR: cmp_rm_keys(): failed to decode input");
+    return -EINVAL;
+  }
+
+  // collect the keys we need to read
+  std::set<std::string> keys;
+  for (const auto& kv : op.entries) {
+    keys.insert(kv.first);
+  }
+
+  // read the values for each key to compare
+  std::map<std::string, bufferlist> values;
+  int r = cls_cxx_map_get_vals_by_keys(hctx, keys, &values);
+  if (r < 0) {
+    CLS_LOG(4, "ERROR: cmp_rm_keys() failed to read values r=%d", r);
+    return r;
+  }
+
+  auto v = values.cbegin();
+  for (const auto& [key, cmp] : op.entries) {
+    if (v == values.end() || v->first != key) {
+      CLS_LOG(20, "cmp_rm_keys() missing key=%s", key.c_str());
+      continue;
+    }
+    CLS_LOG(20, "cmp_rm_keys() comparing key=%s op=%d mode=%d",
+            key.c_str(), (int)cmp.op, (int)cmp.mode);
+
+    const bufferlist& value = v->second;
+    ++v;
+
+    r = compare_value(cmp, value);
+    if (r == -EIO) {
+      r = 0; // treat EIO as a failed comparison
+    }
+    if (r < 0) {
+      CLS_LOG(10, "cmp_rm_keys() failed to compare key=%s r=%d",
+              key.c_str(), r);
+      return r;
+    }
+    if (r == 0) {
+      // unsuccessful comparison
+      CLS_LOG(20, "cmp_rm_keys() preserving key=%s", key.c_str());
+    } else {
+      // successful comparison
+      CLS_LOG(20, "cmp_rm_keys() removing key=%s", key.c_str());
+      r = cls_cxx_map_remove_key(hctx, key);
+      if (r < 0) {
+        CLS_LOG(1, "ERROR: cmp_rm_keys() failed to remove key=%s r=%d",
+                key.c_str(), r);
+        return r;
```

Comment:
maybe better to indicate this could be a partial failure? maybe `-EAGAIN` ?

Reply:
my understanding is that there is no partial failure in rados. if any subop returns an error, then none of earlier subops are committed either

an error from cls_cxx_map_remove_key() here should be fatal, because cls_cxx_map_get_vals_by_keys() just told us the key was there

Reply:
but the other deletions already happened - so it didn't completely failed.
unless there is a way to batch them in one transaction?

Reply:
> batch them in one transaction

the osd does this with the entire osd op. it either all succeeds or all fails

Reply:
ok. got it